src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -173,40 +176,68 @@
+   `java.security.cert.X509Certificate` object.  If the supplied object is nil,
+   the information returned would represent an 'unauthenticated' client."
+  [ssl-client-cert]
+  (if ssl-client-cert
+    (let [cn (ssl-utils/get-cn-from-x509-certificate ssl-client-cert)
+          authenticated (not (empty? cn))]
+      (log/debugf "CN '%s' provided by SSL certificate.  Authenticated: %s."
+                  cn authenticated)
+      {:client-cert-cn cn
+       :authenticated  authenticated})
+    (do
+      (log/debugf "No SSL client certificate provided. "
+                  "Treating client as unauthenticated.")
+      unauthenticated-client-info)))
+
+(defn client-auth-info
+  "Get map of client authentication info for the client.  Map has the following
+  keys:
+
+  * :client-cert - A `java.security.cert.X509Certificate` object or nil
+  * :client-cert-cn - The CN (Common Name) of the client, typically associated
+                      with the CN attribute from the Distinguished Name
+                      in an X.509 certificate's Subject.
+  * :authenticated - A boolean representing whether or not the client is
+                     considered to have been successfully authenticated.
 
-(defn auth-maybe-with-ssl-info
-  "Merge information from the SSL client cert into the jruby request if
-  available and information was not expected to be provided via client headers"
-  [config ssl-client-cert request]
-  (if (:allow-header-cert-info config)
-    request
-    (let [cn   (get-cert-common-name ssl-client-cert)]
-      (merge request {:client-cert    ssl-client-cert
-                      :client-cert-cn cn
-                      :authenticated  (not (nil? cn))}))))
+  Parameters:
+
+  * config - Map of configuration data
+  * request - Ring request containing client data"
+  [config request]
+  (if-let [authorization (:authorization request)]
+    {:client-cert    (ring-auth/authorized-certificate request)
+     :client-cert-cn (ring-auth/authorized-name request)
+     :authenticated  (true? (ring-auth/authorized-authentic? request))}
+    (let [headers (:headers request)
+          header-dn-name (:ssl-client-header config)
+          header-dn-val (get headers header-dn-name)
+          header-auth-name (:ssl-client-verify-header config)
+          header-auth-val (get headers header-auth-name)
+          header-cert-val (get headers header-client-cert-name)]
+      (if (:allow-header-cert-info config)
+        (-> (header-auth-info header-dn-name
+                              header-dn-val
+                              header-auth-name
+                              header-auth-val)
+            (assoc :client-cert (header-cert header-cert-val)))
+        (do
+          (doseq [[header-name header-val]
+                  {header-dn-name header-dn-val
+                   header-auth-name header-auth-val
+                   header-client-cert-name header-cert-val}]
+            (if header-val
```

Comment:
All the comprehension forms have special support for the common conditionals that follow, so you could collapse this `if` into the `doseq` like so:
```clojure
(doseq [[header-name header-val] {...}
        :when header-val]
...)
```

Might be worthwhile in this particular function with as many nesting levels as it has.